### DISCUSSION TITLE
Question about accelerometer and EKF

### DISCUSSION THREAD

**Discussion by wdliu356** [bitcraze/discussions#1506] [Q&A]
Hi all.
I'm working on a customized controller for crazyflie and need the estimation of the speed of the quad. Currently I'm using a flow deck and selecting EKF as the estimator. However, since I'm flying the drone at a very low altitude (not sure whether this is the reason), the estimated speed is not working at all and the estimated speed is completely off. So I'm wondering whether it is possible to use data from accelerometer to estimate the speed and apply that to EKF. Has anyone tried to do that or is it possible to do that? It is also very helpful if anyone can explain why a flow sensor is not working well at a low altitude.

**Comment by knmcguire** [comment]
Do you have a plot that we can look at of the flowdeck's readings or a video of the flight?

We have written a blogpost a while ago with good practises and tips to improve flight with the flow deck, perhaps that might be a good resource for you? https://www.bitcraze.io/2023/11/go-with-the-flow-relative-positioning-with-the-flow-deck/

**Reply by wdliu356** [reply]
I checked the page and I believe there are two reasons for the flow deck not working well. The first one is that currently the floor has very low or almost no texture. The second is the quad is flying at very low altitude and the flow deck is always observing its shadow. I don't think there's a way to solve the second problem so I'm thinking of an alternative way to estimate the speed,. 
As I mentioned in the main post, I'm wondering whether it is possible to add the on-board accelerometer to EKF? A further question is that I noticed it's possible to force the estimator to be EKF even without a flow-deck plugged in. In that case, what sensor is used to estimate the velocity and can you point me where the source code is? I can't find that in either the estimator folder or the kalman core folder.

**Reply by gemenerik** [reply]
If the shadow is the feature that is standing out the most, you might want to consider adding texture to the floor or reducing the shadow hardness by controlling the lighting. This could be a lot easier than modifying the state estimate.

That said, since the firmware is fully open source, you have the flexibility to modify the EKF as needed. If you're interested in integrating additional sensors; the relevant measurements are processed in the firmware, and you can start by looking at how these are handled [here](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/estimator/estimator_kalman.c#L290-L296).
